File: LeetCode/Explore/TopInterview/Amazon/LinkedList/25_reverse_nodes_in_k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  
    def reverseLinkedList(self, head: Optional[ListNode], k: int):
      pre = None
      while k:
        nex = head.next
        head.next = pre
        pre = head
        head = nex
        k -= 1
      return pre
  
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        counter = 0
        pre = head
        
        while counter < k and pre:
          pre = pre.next
          counter += 1
          
        if counter == k:
          reverseHead = self.reverseLinkedList(head, k)
          head.next = self.reverseKGroup(pre, k)
          return reverseHead
        
        return head

File: LeetCode/Explore/TopInterview/Amazon/LinkedList/25_reverse_nodes_in_k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
  
private:
  ListNode* reverseLinkedList(ListNode *head, int k) {
    ListNode *pre = nullptr;
    
    while (k) {
      ListNode *next = head->next;
      head->next = pre;
      pre = head;
      head = next;
      k--;
    }
    return pre;
  }
  
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
      int counter = 0;
      ListNode *pre = head;
      while (counter < k && pre) {
        pre = pre->next;
        counter++;
      }
      if (counter == k) {
        ListNode *reverseHead = reverseLinkedList(head, k);
        head->next = reverseKGroup(pre, k);
        return reverseHead;
      }
      return head;
    }
};
